### PR TITLE
New plugin Color Chooser

### PR DESCRIPTION
**Pull request by micrococo**
A color picker that maintains a list of recently selected colors.

colorchooser_menu

### CODE REVIEW

**Comment by tsujan** [comment]
It seems nice. @palinek reviews panel's codes better that I can.

**Comment by palinek** [comment]
Why didn't you update the existing color picker plugin, but created a new one instead? IMO there is no need to have both color picker and color chooser as the later (color chooser) covers all functionality of the former (color picker).

UX: I definitely would prefer only one icon/button on the panel with left click -> choosing the color (as the "picker" button left click now) and right click -> the context menu (as the "color" button left click now). Also the picked color should IMO be copied into clipboard immediately (to not need another two click to get to it).

**Comment by micrococo** [comment]
> Why didn't you update the existing color picker plugin, but created a new one instead? IMO there is no need to have both color picker and color chooser as the later (color chooser) covers all functionality of the former (color picker).

I didn't know what to do. Some people may prefer the current color picker.
 
> UX: I definitely would prefer only one icon/button on the panel with left click -> choosing the color (as the "picker" button left click now) and right click -> the context menu (as the "color" button left click now).

The plugin is a simplified version of the Color Picker plasmoid from Plasma desktop. I like it as it is now, but I suppose that this is a matter of taste. The right click is left for the usual plugin actions (move, remove, etc.).

> Also the picked color should IMO be copied into clipboard immediately (to not need another two clicks to get to it).

Indeed, and I though I had done it already. I've updated the pull request to include that functionality.

**Comment by palinek** [comment]
> > Why didn't you update the existing color picker plugin, but created a new one instead? IMO there is no need to have both color picker and color chooser as the later (color chooser) covers all functionality of the former (color picker).

> I didn't know what to do. Some people may prefer the current color picker.

@lxqt/lxqt Are there any objections against merging all changes made here into the existing color picker plugin?

**Comment by tsujan** [comment]
No objection on my part.

**Comment by palinek** [comment]
@micrococo please, merge these changes into the color picker plugin.

**Comment by micrococo** [comment]
I've merged the changes into the color picker plugin as requested, @palinek. I hope I haven't messed anything in the way, I'm null with git.

File: plugin-colorpicker/translations/colorpicker_ar.desktop [hunk 1]
```diff
@@ -1,2 +0,0 @@
-Name[ar]=لاقط الألوان
```
Comment (palinek) [inline plugin-colorpicker/translations/colorpicker_ar.desktop]: Why did you remove all those *.desktop translations?

Comment (micrococo) [reply plugin-colorpicker/translations/colorpicker_ar.desktop]: Are they still valid, @palinek? I changed the comment in `colorpicker.desktop.in`.

Comment (palinek) [reply plugin-colorpicker/translations/colorpicker_ar.desktop]: Then, just remove the `Comment` key in localizations...

Comment (micrococo) [reply plugin-colorpicker/translations/colorpicker_ar.desktop]: @palinek, `.desktop` files in the `translations` directory should only contain the `Name[xx]` and `Comment[xx]` keys, shouldn't they? Most of them have the whole content of `colorpicker.desktop.in` plus the localized keys. It is to fix them if it is the case before I update the pull request.

Comment (palinek) [reply plugin-colorpicker/translations/colorpicker_ar.desktop]: > should only contain the Name[xx] and Comment[xx] keys, shouldn't they? 

Yes. But during the build process everything except `Key[lang]` is discarded...

So you're right, you can discard contents copied from original .desktop file.

**Comment by micrococo** [comment]
Pull request updated to include the translations files.

**Comment by micrococo** [comment]
I'm sorry, I forgot to ping you in my last comment, @palinek. The plugin has a serious performance problem. It seems that changing the icon of the `mColorButton` when you pick a new color causes a relayout of the button and this in turn triggers call to the `realign ()` method causing a cycle of calls. I don't know whether this is expected or is something that shouldn't happen. Just to let you know. I have changed the implementation subclassing the button and overwriting the paint method. It seems to work fine but I want to test it better.

File: plugin-colorpicker/colorpicker.cpp [hunk 2]
```diff
@@ -25,81 +25,268 @@
...
+    "     d=\"M 7.3252447,12.939477 C 4.22976,12.490771 2.2993447,9.4012904 3.2660519,6.4430523 3.7960078,4.8213248 5.176118,3.5502752 6.8340753,3.1569853 8.2327213,2.8252076 9.7084288,3.1079402 10.87908,3.9319726 c 0.351893,0.2477004 0.938414,0.8342222 1.186115,1.1861147 0.392347,0.5573834 0.693227,1.2608044 0.828674,1.9373428 0.09455,0.4722872 0.101172,1.3507353 0.01371,1.8196175 -0.12871,0.6899891 -0.443837,1.4407734 -0.842381,2.0069604 -0.247701,0.351892 -0.834222,0.938414 -1.186115,1.186115 -0.549288,0.386648 -1.2303943,0.679168 -1.9132103,0.82168 -0.4104575,0.08567 -1.2226672,0.110259 -1.640625,0.04967 z\""
+    "     id=\"path32\" />"
+    "</svg>"
+};
 
 
-ColorPicker::ColorPicker(const ILXQtPanelPluginStartupInfo &startupInfo) :
-    QObject(),
-    ILXQtPanelPlugin(startupInfo)
+ColorPicker::ColorPicker (const ILXQtPanelPluginStartupInfo &startupInfo) :
```
Comment (palinek) [inline plugin-colorpicker/colorpicker.cpp]: Please, follow the coding style... don't put the space between identifier and opening parenthesis.

**Comment by palinek** [comment]
> changing the icon of the mColorButton when you pick a new color causes a relayout of the button and this in turn triggers

That's a normal behavior in layouts, not? If you change content of one widget (the size could have been changed), the layout of all siblings/parent must act accordingly...

**Comment by micrococo** [comment]
@palinek, I've updated the pull request to fix the problem, but there is a conflict and I don't know what does it means and what do I have to do.

**Comment by palinek** [comment]
> what do I have to do.

Rebase & solve the conflict :)

**Comment by micrococo** [comment]
@palinek, I don't know how to work with git and I messed up the pull request. I think I've fixed it and everything should be fine now.

**Comment by palinek** [comment]
GTM (squashed)

**Comment by tsujan** [comment]
Any plan for reviving this?

**Comment by micrococo** [comment]
> Any plan for reviving this?

Hi, sorry for the late answer. This was already finished by the time of my last comment here and I thought it had been included because palinek gave his GTM. I'm using Plasma and currently I don't have interest in programming. I don't even have development stuff installed, so I can't help with this anymore. Feel free to revoke the request. Sorry for the inconvenience.

**Comment by tsujan** [comment]
@micrococo 
Thanks for your reply!

I might revive it for 0.17.0 if it's OK with you.

**Comment by micrococo** [comment]
> @micrococo
> Thanks for your reply!
> 
> I might revive it for 0.17.0 if it's OK with you.

Of course it is OK. And again, sorry for the inconvenience.

**Comment by palinek** [comment]
Succeeded by #1738

**Comment by tsujan** [comment]
Revived by @palinek in https://github.com/lxqt/lxqt-panel/pull/1738
